lib/srv/tcpip/forward.go

```diff
@@ -0,0 +1,218 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package tcpip
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// ForwardRequest is the payload sent from dialer to forwarder when starting
+// a new forwarding connection.
+type ForwardRequest struct {
+	// DestinationAddress is the target address to dial to.
+	DestinationAddress string `json:"dst_addr"`
+}
+
+// ForwardResponse is the payload sent from forwarder to dialer indicating
+// success/failure in dialing the target address.
+type ForwardResponse struct {
+	// Success is true if the target address was successfully dialed.
+	Success bool `json:"success"`
+	// Error is the error string associated with a failed dial.
+	Error string `json:"error"`
+}
+
+// NewDialer wraps a [utils.SocketPairDialer], sending dials as messages understood by the
+// corresponding [Forwarder].  Using this type in conjunction with
+// the socket pair dialer/listener helpers from utils makes it easy to forward arbitrary
+// TCP dials via a socket.
+func NewDialer(inner utils.SocketPairDialer) *Dialer {
+	return &Dialer{
+		inner: inner,
+	}
+}
+
+// Dialer is a tcp dialer that forwards dials across a socket.
+type Dialer struct {
+	inner utils.SocketPairDialer
+}
+
+// Dial dials the target address.
+func (d *Dialer) Dial(addr string) (net.Conn, error) {
+	conn, err := d.inner.Dial()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	req := ForwardRequest{
+		DestinationAddress: addr,
+	}
```

Comment:
Now that the health-check/ping messages have been removed, yeah.  The custom messages aren't necessary any more.  Since I have these helpers already I like the fact that this system actually lets us return a human-readable error message to the client, but I wouldn't object to switching to using the standard SOCKS5 messages instead.